File: ios/PretextPreparation.swift
import CoreText
import Foundation

internal func prepareInlineParagraphSeed(
    _ paragraph: [InlineSegment],
    baseStyle: NativeTextStyle
) -> NativePreparedParagraphSeed {
    var text = ""
    var tokens: [NativeTokenDescriptor] = []
    var breakUnits: [NativeTokenDescriptor] = []
    var runs: [NativeTextRun] = []
    var atomicSpans: [NativeAtomicSpan] = []
    var inlineBoxes: [NativeInlineBox] = []
    var forceTokenLayout = false

    for segment in paragraph {
        let baseOffset = (text as NSString).length
        let breakBehavior = segment.breakBehavior.lowercased()
        if segment.kind?.lowercased() == inlineSegmentKindBox || segment.boxId != nil {
            let box = buildInlineBox(segment: segment, startUTF16: baseOffset)
            text += objectReplacementCharacter
            runs.append(
                NativeTextRun(
                    startUTF16: box.startUTF16,
                    endUTF16: box.endUTF16,
                    style: baseStyle
                )
            )
            inlineBoxes.append(box)
            atomicSpans.append(
                NativeAtomicSpan(
                    startUTF16: box.startUTF16,
                    endUTF16: box.endUTF16,
                    source: "inline_box"
                )
            )
            tokens.append(
                NativeTokenDescriptor(
                    text: objectReplacementCharacter,
                    startUTF16: box.startUTF16,
                    endUTF16: box.endUTF16,
                    style: baseStyle,
                    inlineBox: box
                )
            )
            breakUnits.append(
                NativeTokenDescriptor(
                    text: objectReplacementCharacter,
                    startUTF16: box.startUTF16,
                    endUTF16: box.endUTF16,
                    style: baseStyle,
                    inlineBox: box
                )
            )
            continue
        }
        let resolvedStyle = resolveTextStyle(segment: segment, baseStyle: baseStyle)
        text += segment.text
        let endOffset = (text as NSString).length
        if endOffset > baseOffset {
            runs.append(
                NativeTextRun(
                    startUTF16: baseOffset,
                    endUTF16: endOffset,
                    style: resolvedStyle
                )
            )
        }
        forceTokenLayout = forceTokenLayout || breakBehavior == breakBehaviorNever
        if breakBehavior == breakBehaviorNever && endOffset > baseOffset {
            atomicSpans.append(
                NativeAtomicSpan(
                    startUTF16: baseOffset,
                    endUTF16: endOffset,
                    source: "inline_break_never"
                )
            )
        }
        appendInlineSegment(
            segment,
            resolvedStyle: resolvedStyle,
            tokens: &tokens,
            breakUnits: &breakUnits
        )
    }

    return NativePreparedParagraphSeed(
        text: text,
        tokens: tokens,
        breakUnits: breakUnits,
        runs: mergeAdjacentRuns(runs),
        atomicSpans: atomicSpans,
        inlineBoxes: inlineBoxes,
        textUnits: text.utf16.count,
        forceTokenLayout: forceTokenLayout
    )
}

private func buildInlineBox(segment: InlineSegment, startUTF16: Int) -> NativeInlineBox {
    let width = max(0, segment.width ?? 0)
    let height = max(0, segment.height ?? 0)
    let baseline = min(max(0, segment.baseline ?? height), height)
    return NativeInlineBox(
        boxId: segment.boxId ?? "inline-box-\(startUTF16)",
        startUTF16: startUTF16,
        endUTF16: startUTF16 + (objectReplacementCharacter as NSString).length,
        width: width,
        height: height,
        baseline: baseline,
        breakBehavior: segment.breakBehavior.lowercased(),
        accessibilityLabel: segment.accessibilityLabel,
        accessibilityHint: segment.accessibilityHint,
        accessibilityRole: segment.accessibilityRole
    )
}

internal func createTypesetter(
    attributedText: NSAttributedString
) -> CTTypesetter {
    return CTTypesetterCreateWithAttributedString(attributedText)
}

internal func prepareMeasuredToken(
    _ token: NativeTokenDescriptor,
    cache: inout [String: NativeTokenMetrics]
) -> NativePreparedToken {
    if let box = token.inlineBox {
        return NativePreparedToken(
            text: token.text,
            startUTF16: token.startUTF16,
            endUTF16: token.endUTF16,
            width: box.width,
            lineHeight: box.height,
            ascent: -box.baseline,
            descent: max(0, box.height - box.baseline)
        )
    }

    if token.text == newlineToken {
        return NativePreparedToken(
            text: token.text,
            startUTF16: token.startUTF16,
            endUTF16: token.endUTF16,
            width: 0,
            lineHeight: 0,
            ascent: 0,
            descent: 0
        )
    }

    let cacheKey = measurementCacheKey(text: token.text, style: token.style)
    let metrics: NativeTokenMetrics
    if let cached = cache[cacheKey] {
        metrics = cached
    } else {
        let measured = measureToken(token.text, style: token.style)
        cache[cacheKey] = measured
        metrics = measured
    }

    return NativePreparedToken(
        text: token.text,
        startUTF16: token.startUTF16,
        endUTF16: token.endUTF16,
        width: metrics.width,
        lineHeight: metrics.lineHeight,
        ascent: metrics.ascent,
        descent: metrics.descent
    )
}

private func measurementCacheKey(
    text: String,
    style: NativeTextStyle
) -> String {
    [
        text,
        style.fontFamily,
        String(style.fontSize),
        String(style.lineHeight),
        String(style.letterSpacing),
        style.locale,
        style.fontWeight,
        style.fontStyle,
        String(style.includeFontPadding),
        style.textDirection.stringValue
    ].joined(separator: "\u{1F}")
}

private func mergeAdjacentRuns(_ runs: [NativeTextRun]) -> [NativeTextRun] {
    guard let first = runs.first else {
        return []
    }

    var merged: [NativeTextRun] = [first]
    for run in runs.dropFirst() {
        let lastIndex = merged.count - 1
        if merged[lastIndex].style == run.style && merged[lastIndex].endUTF16 == run.startUTF16 {
            merged[lastIndex] = NativeTextRun(
                startUTF16: merged[lastIndex].startUTF16,
                endUTF16: run.endUTF16,
                style: run.style
            )
        } else {
            merged.append(run)
        }
    }
    return merged
}

File: android/src/main/java/com/margelo/nitro/pretext/PretextPreparation.kt
package com.margelo.nitro.pretext

import kotlin.math.max

internal fun prepareInlineParagraphSeed(
  paragraph: Array<InlineSegment>,
  baseStyle: NativeTextStyle,
): NativePreparedParagraphSeed {
  val textBuilder = StringBuilder()
  val tokens = ArrayList<NativeTokenDescriptor>()
  val breakUnits = ArrayList<NativeTokenDescriptor>()
  val runs = ArrayList<NativeTextRun>()
  val atomicSpans = ArrayList<NativeAtomicSpan>()
  val inlineBoxes = ArrayList<NativeInlineBox>()
  var forceTokenLayout = false

  paragraph.forEach { segment ->
    val start = textBuilder.length
    val breakBehavior = segment.breakBehavior.lowercase()
    if (segment.kind?.lowercase() == INLINE_SEGMENT_KIND_BOX || segment.boxId != null) {
      val box = buildInlineBox(segment, start)
      textBuilder.append(OBJECT_REPLACEMENT_CHARACTER)
      runs +=
        NativeTextRun(
          start = box.start,
          end = box.end,
          style = baseStyle,
        )
      inlineBoxes += box
      atomicSpans +=
        NativeAtomicSpan(
          start = box.start,
          end = box.end,
          source = "inline_box",
        )
      tokens += NativeTokenDescriptor(
        text = OBJECT_REPLACEMENT_CHARACTER,
        start = box.start,
        end = box.end,
        style = baseStyle,
        inlineBox = box,
      )
      breakUnits += NativeTokenDescriptor(
        text = OBJECT_REPLACEMENT_CHARACTER,
        start = box.start,
        end = box.end,
        style = baseStyle,
        inlineBox = box,
      )
      return@forEach
    }

    val resolvedStyle = resolveTextStyle(segment, baseStyle)
    textBuilder.append(segment.text)
    val end = textBuilder.length
    if (end > start) {
      runs +=
        NativeTextRun(
          start = start,
          end = end,
          style = resolvedStyle,
        )
      if (breakBehavior == BREAK_BEHAVIOR_NEVER) {
        atomicSpans +=
          NativeAtomicSpan(
            start = start,
            end = end,
            source = "inline_break_never",
          )
      }
    }
    forceTokenLayout =
      forceTokenLayout || breakBehavior == BREAK_BEHAVIOR_NEVER
    appendInlineSegmentTokens(
      segment = segment,
      resolvedStyle = resolvedStyle,
      baseOffset = start,
      tokens = tokens,
      breakUnits = breakUnits,
    )
  }

  return NativePreparedParagraphSeed(
    text = textBuilder.toString(),
    tokens = tokens,
    breakUnits = breakUnits,
    runs = mergeAdjacentRuns(runs),
    atomicSpans = atomicSpans,
    inlineBoxes = inlineBoxes,
    textUnits = textBuilder.length,
    forceTokenLayout = forceTokenLayout,
  )
}

private fun buildInlineBox(segment: InlineSegment, start: Int): NativeInlineBox {
  val width = AndroidTextUnits.toPx(max(0.0, segment.width ?: 0.0))
  val height = AndroidTextUnits.toPx(max(0.0, segment.height ?: 0.0))
  val baseline = AndroidTextUnits.toPx(segment.baseline ?: AndroidTextUnits.fromPx(height))
    .coerceIn(0.0, height)
  return NativeInlineBox(
    boxId = segment.boxId ?: "inline-box-$start",
    start = start,
    end = start + OBJECT_REPLACEMENT_CHARACTER.length,
    width = width,
    height = height,
    baseline = baseline,
    breakBehavior = segment.breakBehavior.lowercase(),
    accessibilityLabel = segment.accessibilityLabel,
    accessibilityHint = segment.accessibilityHint,
    accessibilityRole = segment.accessibilityRole,
  )
}

internal fun prepareMeasuredToken(
  token: NativeTokenDescriptor,
  cache: MutableMap<String, NativeTokenMetrics>,
): NativePreparedToken {
  token.inlineBox?.let { box ->
    return NativePreparedToken(
      text = token.text,
      start = token.start,
      end = token.end,
      width = box.width,
      lineHeight = box.height,
      ascent = -box.baseline,
      descent = box.height - box.baseline,
    )
  }

  if (token.text == NEWLINE_TOKEN) {
    return NativePreparedToken(
      text = token.text,
      start = token.start,
      end = token.end,
      width = 0.0,
      lineHeight = 0.0,
      ascent = 0.0,
      descent = 0.0,
    )
  }

  val cacheKey = measurementCacheKey(token.text, token.style)
  val metrics = cache.getOrPut(cacheKey) { measureToken(token.text, token.style) }
  return NativePreparedToken(
    text = token.text,
    start = token.start,
    end = token.end,
    width = metrics.width,
    lineHeight = metrics.lineHeight,
    ascent = metrics.ascent,
    descent = metrics.descent,
  )
}

private fun measurementCacheKey(text: String, style: NativeTextStyle): String {
  return listOf(
    text,
    style.fontFamily,
    style.fontSize.toString(),
    style.lineHeight.toString(),
    style.letterSpacing.toString(),
    style.locale,
    style.fontWeight,
    style.fontStyle,
    style.includeFontPadding.toString(),
    style.textDirection.name,
  ).joinToString(separator = "\u001F")
}

private fun mergeAdjacentRuns(runs: List<NativeTextRun>): List<NativeTextRun> {
  if (runs.isEmpty()) {
    return emptyList()
  }

  val merged = ArrayList<NativeTextRun>(runs.size)
  merged += runs.first()
  for (index in 1 until runs.size) {
    val run = runs[index]
    val previous = merged.last()
    if (previous.style == run.style && previous.end == run.start) {
      merged[merged.lastIndex] =
        NativeTextRun(
          start = previous.start,
          end = run.end,
          style = run.style,
        )
    } else {
      merged += run
    }
  }
  return merged
}
